src/ApparentHorizons/StrahlkorperGr.cpp

```diff
@@ -6,15 +6,354 @@
+
+    // \nabla R \cdot \nabla Y_lm
+    auto grad_ricci_scalar_dot_grad_yi =
+        make_with_value<Scalar<DataVector>>(ricci_scalar, 0.0);
+    get(grad_ricci_scalar_dot_grad_yi) += get<0>(derivs_yi.first) *
+                                          get<0>(grad_ricci_scalar) *
+                                          get<0, 0>(inverse_surface_metric);
+    get(grad_ricci_scalar_dot_grad_yi) +=
+        get<0>(derivs_yi.first) * get<1>(grad_ricci_scalar) *
+        get<1, 0>(inverse_surface_metric) * get(sin_theta);
+    get(grad_ricci_scalar_dot_grad_yi) +=
+        get<1>(derivs_yi.first) * get<0>(grad_ricci_scalar) *
+        get<1, 0>(inverse_surface_metric) * get(sin_theta);
+    get(grad_ricci_scalar_dot_grad_yi) +=
+        get<1>(derivs_yi.first) * get<1>(grad_ricci_scalar) *
+        get<1, 1>(inverse_surface_metric) * square(get(sin_theta));
+
+    // Assemble the operator making up the eigenproblem's left-hand-side
+    auto left_matrix_yi_physical =
+        make_with_value<Scalar<DataVector>>(ricci_scalar, 0.0);
+    get(left_matrix_yi_physical) = get(laplacian_squared_yi) +
+                                   get(ricci_scalar) * get(laplacian_yi) +
+                                   get(grad_ricci_scalar_dot_grad_yi);
+
+    // Transform back to spectral space, to get one column each for the left
+    // and right matrices for the eigenvalue problem.
+    const DataVector left_matrix_yi_spectral =
+        ylm.phys_to_spec(get(left_matrix_yi_physical));
+    const DataVector right_matrix_yi_spectral =
+        ylm.phys_to_spec(get(laplacian_yi));
+
+    // Set the current column of the left and right matrices
+    // for the eigenproblem.
+    if (iter_i.l() > 0 and iter_i.l() < ylm.l_max() - 1) {
+      size_t row = 0;
+      for (iter_j.reset(); iter_j; ++iter_j) {
+        if (iter_j.l() > 0 and iter_j.l() < ylm.l_max() - 1) {
+          (*left_matrix)(row, column) = left_matrix_yi_spectral[iter_j()];
+          (*right_matrix)(row, column) = right_matrix_yi_spectral[iter_j()];
+          ++row;
+        }
+      }  // loop over rows
+      ++column;
+    }
+  }  // loop over columns
+}
+
+// Find the eigenvectors corresponding to the three smallest-magnitude
+// eigenvalues.
+// Note: uses the fact that eigenvalues should be real
+std::array<DataVector, 3> get_eigenvectors_for_3_smallest_magnitude_eigenvalues(
+    const DataVector& eigenvalues_real_part, const Matrix& eigenvectors,
+    const YlmSpherepack& ylm) noexcept {
+  double eigenvalue_smallest = max(abs(eigenvalues_real_part));
+  double eigenvalue_second_smallest = max(abs(eigenvalues_real_part));
+  double eigenvalue_third_smallest = max(abs(eigenvalues_real_part));
+  size_t index_smallest = 0;
+  size_t index_second_smallest = 0;
+  size_t index_third_smallest = 0;
+
+  // Simple algorithm that loops over all elements
+  // Improve with quick select algorithm if necessary
+  for (size_t i = 0; i < eigenvalues_real_part.size(); ++i) {
```

Comment:
I have no idea what this code is doing (it's long and confusing with lots of stuff happening) but if all you want are the indices of the three smallest eigenvectors:
```
template <size_t K, typename T>
auto k_smallest_index(const T& t) -> std::array<size_t, K> {
  std::array<size_t, K> temp{};
  std::iota(temp.begin(), temp.end(), 0);
  const auto get_max_index = [&temp, &t]() {
    return std::distance(
        temp.begin(),
        std::max_element(temp.begin(), temp.end(),
                         [&t](const size_t lhs, const size_t rhs) {
                           return t[lhs] < t[rhs];
                         }));
  };
  size_t max_index = get_max_index();
  for (size_t i = K; i < t.size(); ++i) {
    if (t[i] < t[max_index]) {
      temp[max_index] = i;
      max_index = get_max_index();
    }
  }
  return temp;
}
```
To get it sorted by the eigenvectors after you can use:
```
  std::vector<double> t{-1.0, 7.8, 9.3, -22.8, 9.5};
  auto result = k_smallest_index<3>(t);
  std::sort(
      result.begin(), result.end(),
      [&t](const size_t lhs, const size_t rhs) { return t[lhs] < t[rhs]; });
  for (const auto& p : result) {
    std::cout << t[p] << "\n";
  }
```
depending on what you want you'll need to add absolute values in places. The advantage of the above approach is that it uses STL algorithms and so is quite easy to reason about. The `k_smallest_indices` should probably be generalized to take a comparator as an argument and then it would be useful in `Utilities/Algorithm.hpp`.

Reply:
After discussing with @fmahebert and @markscheel, I've cleaned up my algorithm to find the top 3 smallest eigenvalues. It should now be straightforward for you to follow it, @nilsdeppe . Here it is (note I need the smallest magnitude, not just the smallest):

```
  size_t index_smallest = 0;
  size_t index_second_smallest = 0;
  size_t index_third_smallest = 0;

  // Simple algorithm that loops over all elements to
  // find indexes of 3 smallest-magnitude eigenvalues
  for (size_t i = 1; i < eigenvalues_real_part.size(); ++i) {
    if (abs(eigenvalues_real_part[i]) <
        abs(eigenvalues_real_part[index_smallest])) {
      index_third_smallest = index_second_smallest;
      index_second_smallest = index_smallest;
      index_smallest = i;
    } else if (i < 2 or abs(eigenvalues_real_part[i]) <
               abs(eigenvalues_real_part[index_second_smallest])) {
      index_third_smallest = index_second_smallest;
      index_second_smallest = i;
    } else if (i < 3 or abs(eigenvalues_real_part[i]) <
               abs(eigenvalues_real_part[index_third_smallest])) {
      index_third_smallest = i;
    }
  }
```

You are welcome to submit a separate PR that implements a more generic "find n smallest elements in a container," @nilsdeppe , along the lines you suggested above.

Reply:
I fail to see how this is easier than taking the code I suggested and copying it into `Utilities` somewhere, but that's besides the point. I'm not have of the complex if/else logic.

Reply:
@nilsdeppe: As you are objecting to the current implementation because of its "complexity", I think it may help if you could define what type of complexity you are talking about. I don't think you mean formal O(N) complexity, because both algorithms look equivalent in this regard. Anecdotally, I found that I could easily make sense of @geoffrey4444's algorithm because it follows how I intuitively think about inserting a new small value into the previous three smallest values. In contrast, I found the algorithm you suggested to be more burdensome to make sense of, due to the more deeply nested, but undocumented, code flow.

I definitely favor adding a generic, re-usable function; however, I think it would be helpful if you could improve the readability of your algorithm via some documentation or more precise variable/function names (suggestions: improve distinction between indices into `t` vs. indices into `temp`, `max_index` -> `index_of_max_value`, and so on).

Reply:
The "simple" case has way more code logic flow, whereas what I presented only uses STL algorithms which are easy to reason about. I'm not editing my comment to change variable names, that's not an efficient use of my time.

Reply:
I like @geoffrey4444's latest version, although it should use arrays instead of prefixing/suffixing all the variables with "smallest", "second_smallest", and "third_smallest".  I think if you initialize the indices to 0,1,2 you can also avoid the `i < 2` and `i < 3` in the later if checks, but I haven't examined that in detail.

The if-chain could be replaced by a loop over an if, but for three elements I'm not sure it would be better than the unrolled form.

Reply:
I’ll fix this to use arrays, @wthrowe...thanks! I guess I thought an array allocation might have more overhead than a few doubles.

I’m fine changing to an if loop, but I thought this was clearer.

I thought about your idea already, but I figured you could run into trouble without the I<2 and i<3 checks, since otherwise you must always have the 3 elements correctly sorted. Eg suppose the first to third smallest elements are actually 2,0,1. Then your proposed scheme would give 0,1,2 -> 0,0,1 -> 0,0,1 -> 2,0,0.

@nilsdeppe, I appreciate your concern about time efficiency and about preferring prerolled solutions in STL where you can. I just feel it’s not an efficient use of my time to switch away from an algorithm that is already correct, O(N), clear to read, and sufficient for our science needs for the moment.

Reply:
See #818